cli/replicaset/cmd/roles.go

```diff
@@ -0,0 +1,121 @@
+package replicasetcmd
+
+import (
+	"fmt"
+
+	"github.com/apex/log"
+	"github.com/tarantool/tt/cli/connector"
+	"github.com/tarantool/tt/cli/replicaset"
+	"github.com/tarantool/tt/cli/running"
+	libcluster "github.com/tarantool/tt/lib/cluster"
+)
+
+// RolesAddCtx describes the context to add role to
+// provided config scope.
+type RolesAddCtx struct {
+	// InstName is an instance name in which add or remove role.
+	InstName string
+	// GroupName is an replicaset name in which add or remove role.
+	GroupName string
+	// ReplicasetName is an replicaset name in which add or remove role.
+	ReplicasetName string
+	// IsGlobal is an boolean value if role needs to add in global scope.
+	IsGlobal bool
+	// RoleName is a name of role to add.
+	RoleName string
+	// Publishers is data publisher factory.
+	Publishers libcluster.DataPublisherFactory
+	// Collectors is data collector factory.
+	Collectors libcluster.DataCollectorFactory
+	// IsApplication true if an application passed.
+	IsApplication bool
+	// Orchestrator is a forced orchestator choice.
+	Orchestrator replicaset.Orchestrator
+	// Conn is an active connection to a passed instance.
+	Conn connector.Connector
+	// RunningCtx is an application running context.
+	RunningCtx running.RunningCtx
+	// Force true if unfound instances can be skipped.
+	Force bool
+	// Timeout describes a timeout in seconds.
+	// We keep int as it can be passed to the target instance.
+	Timeout int
+}
+
+// RolesAdd adds role with provided path target to config.
+func RolesAdd(ctx RolesAddCtx) error {
+	var (
+		orchestratorType replicaset.Orchestrator
+		err              error
+	)
+	orchestratorType, err = getInstanceOrchestrator(ctx.Orchestrator, ctx.Conn)
+	if err != nil {
+		return err
+	}
+
+	if orchestratorType == replicaset.OrchestratorCartridge {
+		if ctx.ReplicasetName == "" {
+			return fmt.Errorf("in cartridge need to specify replica set name via --replicaset flag")
+		}
+	}
+
+	var orchestrator replicasetOrchestrator
+	if ctx.IsApplication {
+		if orchestrator, err = makeApplicationOrchestrator(
+			orchestratorType, ctx.RunningCtx, ctx.Collectors, ctx.Publishers); err != nil {
+			return err
+		}
+	} else {
+		if orchestrator, err = makeInstanceOrchestrator(orchestratorType, ctx.Conn); err != nil {
+			return err
+		}
+	}
+
+	log.Info("Discovery application...")
+	fmt.Println()
+
+	// Get and print status.
+	replicasets, err := orchestrator.Discovery(replicaset.SkipCache)
+	if err != nil {
+		return err
+	}
+	statusReplicasets(replicasets)
+	fmt.Println()
+
+	if ctx.IsGlobal {
+		if orchestratorType == replicaset.OrchestratorCartridge {
+			log.Warn("Ignoring --global (-G) flag due to cluster with cartridge" +
+				" orchestrator can't add role globally")
+		} else {
+			log.Infof("Add role %s to global scope", ctx.RoleName)
+		}
+	}
+	if ctx.GroupName != "" && orchestratorType != replicaset.OrchestratorCartridge {
+		log.Infof("Add role %s to group: %s", ctx.RoleName, ctx.GroupName)
+	}
```

Comment:
Why is there no message in the case of a Cartridge?

Reply:
In case of Cartridge orchestrator we have VShardGroup that can be added.